ionoscloud/resource_server.go

```diff
@@ -812,6 +869,107 @@ func resourceServerUpdate(ctx context.Context, d *schema.ResourceData, meta inte
+		return diags
+	}
+
+	if d.HasChange("vm_state") {
+		_, newState := d.GetChange("vm_state")
+		newVmState := newState.(string)
+		switch *serverType {
+		case "ENTERPRISE":
+			if newVmState == "SUSPENDED" {
+				diags := diag.FromErr(fmt.Errorf("cannot suspend an enterprise server, set to SHUTOFF instead"))
+				return diags
+			}
+			if newVmState == "RUNNING" && *currentVmState == "SHUTOFF" {
+				apiResponse, err := client.ServersApi.DatacentersServersStartPost(ctx, dcId, d.Id()).Execute()
+				logApiRequestTime(apiResponse)
+				if err != nil {
+					return diag.FromErr(err)
+				}
+				err = utils.WaitForResourceToBeReady(ctx, d, func(ctx context.Context, d *schema.ResourceData) (bool, error) {
+					ionoscloudServer, _, err := client.ServersApi.DatacentersServersFindById(ctx, dcId, d.Id()).Execute()
+					if err != nil {
+						return false, err
+					}
+					if *ionoscloudServer.Properties.VmState == *currentVmState {
+						log.Printf("[INFO] Enterprise server vmState not yet changed to RUNNING: %s", d.Id())
+						return false, nil
+					}
+					return true, nil
+				})
+
+				if err != nil {
+					return diag.FromErr(err)
+				}
+			}
+			if newVmState == "SHUTOFF" && *currentVmState == "RUNNING" {
+				apiResponse, err := client.ServersApi.DatacentersServersStopPost(ctx, dcId, d.Id()).Execute()
+				logApiRequestTime(apiResponse)
+				if err != nil {
+					return diag.FromErr(err)
+				}
+				err = utils.WaitForResourceToBeReady(ctx, d, func(ctx context.Context, d *schema.ResourceData) (bool, error) {
+					ionoscloudServer, _, err := client.ServersApi.DatacentersServersFindById(ctx, dcId, d.Id()).Execute()
+					if err != nil {
+						return false, err
+					}
+					if *ionoscloudServer.Properties.VmState == *currentVmState {
+						log.Printf("[INFO] Enterprise server vmState not yet changed to SHUTOFF: %s", d.Id())
+						return false, nil
+					}
+					return true, nil
+				})
+
+				if err != nil {
+					return diag.FromErr(err)
+				}
+			}
+		case "CUBE":
+			if newVmState == "SHUTOFF" {
+				diags := diag.FromErr(fmt.Errorf("cannot shut down a cube server, set to SUSPENDED instead"))
+				return diags
+			}
+			if newVmState == "" {
+				diags := diag.FromErr(fmt.Errorf("cannot update a suspended Cube Server, must change the state to RUNNING first"))
+				return diags
+			}
+			// Unsuspend a Cube server first, before applying other changes
+			if newVmState == "RUNNING" && *currentVmState == "SUSPENDED" {
+				apiResponse, err := client.ServersApi.DatacentersServersResumePost(ctx, dcId, d.Id()).Execute()
+				logApiRequestTime(apiResponse)
+				if err != nil {
+					return diag.FromErr(err)
+				}
+
+				err = utils.WaitForResourceToBeReady(ctx, d, func(ctx context.Context, d *schema.ResourceData) (bool, error) {
+					ionoscloudServer, _, err := client.ServersApi.DatacentersServersFindById(ctx, dcId, d.Id()).Execute()
```

Comment:
can we have these inline functions to normal function? They seem very similar

Reply:
refactored